tokio-util/src/sync/cancellation_token/implementation.rs

```diff
@@ -0,0 +1,341 @@
+    let mut locked_node = node.inner.lock().unwrap();
+
+    // Once no handles are left over, the node gets detached from the tree.
+    // There should never be a new handle once all handles are dropped.
+    assert!(locked_node.num_handles > 0);
+
+    locked_node.num_handles += 1;
+}
+
+/// Decreses the reference count of handles.
+///
+/// Once no handle is left, we can remove the node from the
+/// tree and connect its parent directly to its children.
+pub(crate) fn decrease_handle_refcount(node: &Arc<TreeNode>) {
+    let num_handles = {
+        let mut locked_node = node.inner.lock().unwrap();
+        locked_node.num_handles -= 1;
+        locked_node.num_handles
+    };
+
+    if num_handles == 0 {
+        with_locked_node_and_parent(node, |mut node, parent| {
+            // Remove the node from the tree
+            match parent {
+                Some(mut parent) => {
+                    // As we want to remove ourselves from the tree,
+                    // we have to move the children to the parent, so that
+                    // they still receive the cancellation event without us.
+                    // Moving them does not violate invariant #1.
+                    move_children_to_parent(&mut node, &mut parent);
+
+                    // Remove the node from the parent
+                    remove_child(&mut parent, node);
+                }
+                None => {
+                    // Due to invariant #1, we can assume that our
+                    // children can no longer be cancelled through us.
+                    // (as we now have neither a parent nor handles)
+                    // Therefore we can disconnect them.
+                    disconnect_children(&mut node);
+                }
+            }
+        });
+    }
+}
+
+/// Cancels a node and its children.
+///
+/// Then, dissolves the entire subtree for easier disposal.
+///
+/// Invariant #1:
+/// It's important that during the entire process of cancellation, there is
+/// never a gap in lock propagation. We need to lock both the parent and the child for the
+/// entire duration of cancellation and disconnection. Those cannot happen separately,
+/// otherwise other calls like [decrease_handle_refcount] could modify the tree structure
+/// while we recurse over it.
+///
+/// We basically have to create a "lock wave" from the parent to children, that nothing else
+/// can bypass. Otherwise, other actions could add/remove children that our cancellation would miss.
+///
+/// Example: Holding a reference of a child does not influence its handle reference count, so
+/// it will happily drop its grandchildren in [decrease_handle_refcount] before we got the chance
+/// to propagate the cancellation to them, if we violate invariant #1.
+/// Unless we keep the child locked during the entire time, from the moment on when we
+/// disconnected it from its parent.
+pub(crate) fn cancel(node: &Arc<TreeNode>) {
+    let waker = &node.waker;
```

Comment:
You should wake the wakers after releasing the lock. Otherwise the future might try to lock the mutex while we're still holding it, which is a waste of resources.

Reply:
Fixed